src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs

```diff
@@ -192,39 +198,38 @@ private void CommonInit()
                 }
                 else
                 {
-                    SearchResult searchResult = CreateSearchResult(_searchData, data);
-                    if (_resultHandler.IsResultIncluded(searchResult))
+                    TSource result;
+                    if (IsResultIncluded(ref data, out result))
                     {
-                        current = _resultHandler.CreateObject(searchResult);
+                        current = result;
                     }
                 }
             }
-            // for AllDirectories, we first recurse into dirs, so cleanup and add searchData 
+            // for AllDirectories, we first recurse into dirs, so cleanup and add searchData
             // to the stack
             else
             {
                 _hnd.Dispose();
+                _searchStack = new List<SearchData>();
```

Comment:
I'm going to go ahead and submit a PR for this by adding an internal `Array.Reverse<T>` and making `List<T>.Reverse` use it.